src/anml/data/data.py

```diff
@@ -168,21 +176,51 @@ def process_params(self, df: pd.DataFrame):
         
         design_mat_blocks = []
         constr_mat_blocks = []
+        re_mat_groups = defaultdict(dict)
         lbs = []
         ubs = []
+        self.fe_variables_names = []
         for param_set in self._param_set:
             for parameter in param_set.parameters:
                 for variable in parameter.variables:
-                    design_mat_blocks.append(variable.design_mat(df=df))
+                    design_mat = variable.design_mat(df=df)
+                    design_mat_blocks.append(design_mat)
+                    var_name = parameter.param_name + '_' + variable.covariate
+                    self.fe_variables_names.append(var_name)
+                    if variable.add_re:
+                        re_mat_groups[variable.col_group][var_name] = design_mat
                     mat, lb, ub = variable.get_constraint_matrix()
                     constr_mat_blocks.append(mat)
                     lbs.append(lb)
                     ubs.append(ub)
-        design_matrix = np.hstack(design_mat_blocks)
-        constr_matrix = block_diag(*constr_mat_blocks)
-        lower_bounds = np.hstack(lbs)
-        upper_bounds = np.hstack(ubs)
-        assert design_matrix.shape[1] == constr_matrix.shape[1]
-        assert len(lower_bounds) == len(upper_bounds) == constr_matrix.shape[0]
-
-        return design_matrix, constr_matrix, lower_bounds, upper_bounds
+        self.design_matrix = np.hstack(design_mat_blocks)
+        self.constr_matrix = block_diag(*constr_mat_blocks)
+        self.constr_lower_bounds = np.hstack(lbs)
+        self.constr_upper_bounds = np.hstack(ubs)
+        assert self.design_matrix.shape[1] == self.constr_matrix.shape[1]
+        assert len(self.constr_lower_bounds) == len(self.constr_upper_bounds) == self.constr_matrix.shape[0]
+
+        if len(re_mat_groups) == 0:
+            self.re_matrix = None
+        else:
+            self.groups_info = defaultdict(dict)
+            re_mat_blocks = []
+            self.re_variables_names = []
+            for col_group, dct in re_mat_groups.items():
+                self.encode_groups(col_group, df)
+                grp_assign = [self.groups_info[col_group][g] for g in df[col_group]]
+                n_group = len(self.groups_info[col_group])
+
+                self.re_variables_names.append(dct.keys())
```

Comment:
If `re_variables_names` can be a `list` of `list` instead of a `list` of dictionary keys that will be great.

Reply:
thanks!